serde-support/src/main/java/io/micronaut/serde/support/deserializers/DemuxingObjectDecoder.java

```diff
@@ -0,0 +1,281 @@
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.serde.support.deserializers;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.type.Argument;
+import io.micronaut.serde.Decoder;
+import io.micronaut.serde.DelegatingDecoder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Special decoder that <i>demuxes</i> an object: The same object can be iterated over multiple
+ * times with multiple decoders (concurrently or sequentially, though multi-threading is not
+ * allowed). Properties that are skipped by one decoder can be read by another.
+ * <br>
+ * The use for this is subtype detection: An object is iterated over once to find the {@code type}
+ * property to detect the subtype, and then the remaining properties (before and after the type)
+ * are deserialized as normal.
+ *
+ * @author Jonas Konrad
+ * @since 2.2.7
+ */
+@Internal
+final class DemuxingObjectDecoder extends DelegatingDecoder {
+    private final DemuxerState state;
+    private int nextKeyIndex;
+
+    private DemuxingObjectDecoder(DemuxerState state, int nextKeyIndex) {
+        this.state = state;
+        this.nextKeyIndex = nextKeyIndex;
+        state.outputCount++;
+    }
+
+    /**
+     * Create a new <i>primed</i> decoder that can decode the same object multiple times. This
+     * decoder is very restricted: It <i>must</i> be {@link AutoCloseable#close() closed} after
+     * use, and it <i>only</i> supports {@link #decodeObject()}. Each {@link #decodeObject()} call
+     * returns a decoder of the same object.
+     *
+     * <pre>{@code
+     * try (Decoder primed = DemuxingObjectDecoder.prime(...)) {
+     *     Decoder d1 = primed.decodeObject();
+     *     decodeSomeProperties(d1);
+     *     d1.finishStructure(true);
+     *
+     *     Decoder d2 = primed.decodeObject();
+     *     decodeOtherProperties(d2);
+     *     d2.finishStructure(true);
+     * }
+     * }</pre>
+     *
+     * @param decoder The input to read from. The primed decoder will call {@link #decodeObject()}
+     *                on this input exactly once
+     * @return The primed decoder
+     */
+    public static Decoder prime(Decoder decoder) {
+        return new PrimedDecoder(decoder);
+    }
+
+    @Override
+    public @Nullable String decodeKey() throws IOException {
+        DemuxerState.Entry entry;
+        do {
+            entry = state.getEntry(nextKeyIndex++);
```

Comment:
@yawkat Can this be rewritten using an iterator?

Reply:
no because the list may change size